idx/memory/tag_query.go

```diff
@@ -0,0 +1,176 @@
+package memory
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/raintank/metrictank/idx"
+)
+
+var (
+	expressionRe    = regexp.MustCompile("^([^!=~]+)([!=~]{1,3})(.*)$")
+	errUnknownId    = errors.New("id does not exist")
+	errInvalidTag   = errors.New("found an invalid tag in the index")
+	errInvalidQuery = errors.New("invalid query")
+)
+
+type TagQuery struct {
+	selects []func(TagIndex) (map[string]struct{}, error)
+	filters []func(def *idx.Archive) (bool, error)
+}
+
+func NewTagQuery(expressions []string) (*TagQuery, error) {
+	if len(expressions) == 0 {
+		return nil, errInvalidQuery
+	}
+
+	query := &TagQuery{}
+
+	for _, expr := range expressions {
+		if !expressionRe.MatchString(expr) {
```

Comment:
a little hand written loop to check the format should easily outperform this

Reply:
I added custom expression parsing now. But i'm not sure if that's not overcomplicating it: https://github.com/grafana/metrictank/pull/729/commits/ac79a31965b0844a822256f313ad6e028fccac7f

Reply:
i suggest:
1) collect cpu profile while running an tag search benchmark, using the regex parsing approach
2) seeing if the expression part is significant. if not, you can stick to the regex approach
3) if it is, do new benchmark using the new code and see how much is saved. (you can also isolate out the 2 parsing approaches into a function, and benchmark those, to get a direct comparison)

Reply:
After I started benchmarking I realized that the performance was by far not good enough and started optimizing the way how the intersects are done. With the commit https://github.com/grafana/metrictank/pull/729/commits/e59f11b638816c7cef99aa967623266954314329 the performance has improved a lot:

```
before e59f11b638816c7cef99aa967623266954314329

pkg: github.com/grafana/metrictank/idx/memory
BenchmarkTagFindSimpleIntersect-8   	1	35819673182 ns/op	21574880 B/op	   10427 allocs/op
PASS
ok  	github.com/grafana/metrictank/idx/memory	67.447s

after e59f11b638816c7cef99aa967623266954314329

pkg: github.com/grafana/metrictank/idx/memory
BenchmarkTagFindSimpleIntersect-8   	10	110105738 ns/op	12124306 B/op	    5728 allocs/op
PASS
ok  	github.com/grafana/metrictank/idx/memory	45.196s
```

Then I wanted to know if switching it back to using the regex library would slow it down, so I switched it back to the regex library while leaving the other improvements in the code and that's how they compare:

```
with custom expression parsing:

pkg: github.com/grafana/metrictank/idx/memory
BenchmarkTagFindSimpleIntersect-8   	10	110105738 ns/op	12124306 B/op	    5728 allocs/op
PASS
ok  	github.com/grafana/metrictank/idx/memory	45.196s

with re library expression parsing:

pkg: github.com/grafana/metrictank/idx/memory
BenchmarkTagFindSimpleIntersect-8   	10	 122717274 ns/op	12129276 B/op	    5749 allocs/op
PASS
ok  	github.com/grafana/metrictank/idx/memory	44.327s
```

Reply:
I added that benchmark to isolate the expression parsing: https://github.com/grafana/metrictank/pull/729/commits/c0eade95b4dd241e751ab4526e899a627db85ff9

When running it with the regular expression library based implementation it seems to perform clearly worse:

```
with re:
goos: linux
goarch: amd64
pkg: github.com/grafana/metrictank/idx/memory
BenchmarkExpressionParsing-8   	  100000	     12555 ns/op	    3528 B/op	      52 allocs/op
PASS
ok  	github.com/grafana/metrictank/idx/memory	25.473s

with custom parsing:
goos: linux
goarch: amd64
pkg: github.com/grafana/metrictank/idx/memory
BenchmarkExpressionParsing-8   	 1000000	      1819 ns/op	     664 B/op	      50 allocs/op
PASS
ok  	github.com/grafana/metrictank/idx/memory	29.139s
```

Reply:
nice